packages/agent/src/pid.ts

```diff
@@ -0,0 +1,139 @@
+ * @returns The path to the PID file
+ */
+export function getPidFilePath(appName: string): string {
+  switch (platform()) {
+    case 'linux':
+    case 'darwin':
+      // We use tmpdir for linux and Mac because /var/run requires root access
+      // The benefit of using the convention is outweighed by the required permission level to create files/directories inside
+      return path.join(tmpdir(), 'medplum-agent', `${appName}.pid`);
+    case 'win32':
+      return path.join('C:', 'ProgramData', 'MedplumAgent', 'pids', `${appName}.pid`);
+    default:
+      throw new Error('Invalid OS');
+  }
+}
+
+/**
+ * Remove the PID file
+ * @param pidFilePath - Path to the PID file
+ */
+export function removePidFile(pidFilePath: string): void {
+  if (fs.existsSync(pidFilePath)) {
+    try {
+      fs.unlinkSync(pidFilePath);
+      pidLogger.info(`PID file removed: ${pidFilePath}`);
+    } catch (err) {
+      pidLogger.error(`Error removing PID file: ${pidFilePath}`, err as Error);
+    }
+  }
+}
+
+/**
+ * Returns true if a PID points to an existing process, else returns false.
+ * @param pid - The PID to check if it exists.
+ * @returns Boolean indicating if a process with the given PID exists.
+ */
+export function checkProcessExists(pid: number): boolean {
+  try {
+    // Sending signal 0 doesn't actually send a signal but checks if process exists
+    process.kill(pid, 0);
+    // If we didn't throw an error above, the process is still running, and PID is not stale
+    return true;
+  } catch (err) {
+    // If the error is "ESRCH", the process doesn't exist
+    if ((err as Error & { code: string }).code === 'ESRCH') {
+      return false;
+    }
+    // If the error is "EPERM", the process exists but we don't have permission
+    if ((err as Error & { code: string }).code === 'EPERM') {
+      return true;
+    }
+    // For any other errors, throw them
+    throw err;
+  }
+}
+
+/**
+ * Create a PID file for the current process
+ * @param appName - Optional application name, defaults to script filename
+ * @returns Object containing success status and pidFilePath
+ */
+export function createPidFile(appName: string): string {
+  const pid = process.pid;
+  const pidFilePath = getPidFilePath(appName);
+
+  // Check if PID file already exists
+  if (fs.existsSync(pidFilePath)) {
+    const existingPid = fs.readFileSync(pidFilePath, 'utf8').trim();
+
+    // Check if the process with this PID is still running
+    pidLogger.info('Checking if process is running');
+    if (checkProcessExists(Number.parseInt(existingPid, 10))) {
+      throw new Error(`${appName} already running`);
+    }
+
+    // If we make it here, PID file exists but it's stale
+    pidLogger.info('Stale PID file found. Overwriting...');
+  }
+
+  // We need to make sure the directory exists first since we aren't just using a system-created dir like /var/run
+  ensureDirectoryExists(path.dirname(pidFilePath));
+
+  // Write the PID file atomically using a temporary file
```

Comment:
Can you expand on why the PID file needs to be written atomically and how using the `.tmp` suffix ensures that? It seems like it'd need to be a random extension if this is geared at competing medplum agent processes attempting to write the PID file? In other words, doesn't this just shift the potential for conflict from `pidFilePath` to `${pidFilePath}.tmp`?

Reply:
That's a good callout, yeah competing agents was the thought but you're right, technically the second process could clobber the PID in the tmp file but it wouldn't be able to rename the file since it won't exist after the first process renamed it. So it could be in a bad state.

I looked at the Node docs again and remembered that using the `wx` flags should give us the atomic write to prevent the PID file from being overwritten, so I'll change it to use that 👍